newsrevenuehub/rev-engine
# DEV-3619: Extract payment-related logic to hook

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

#### What's this PR do?

The initial goal of the PR was to prevent billing country from being blank. However, because it was originally in shared context, it was nearly impossible to do this.

The new way this works:

1. The user visits a contribution page and completes the form.
2. The page creates a Payment object (using a new hook) which contains all of the information entered in the form along with info provided when we POST to `/payments` to create a new Stripe payment. The hook validates that billing country is not blank, among other things.
3. The page opens a FinishPaymentModal component with this payment object. Because the object can't be changed directly by the page, we know the country remains valid. (Before this PR, it was possible for browser autofill to make changes to the form _under_ the Stripe modal and potentially trigger changes to the payment information.)
4. The FinishPaymentModal makes Stripe calls to finalize the payment based on the info provided by the user.
5. If the user cancels out of FinishPaymentModal, the payment is deleted and we start from the top.

Other related changes:
- Removed coupling between EditRecurringPaymentModal styles and StripePaymentForm.
- Removed coupling between DonationPage and StripePaymentWrapper. StripePaymentWrapper can now be used wherever we need it to be; it takes needed info as props instead of looking for a donation page context.
- Removed properties from DonationPage context that are now stored in the Payment object.
- Migrated DonationPageDisclaimer and StripePaymentForm to TypeScript.
- Corrected type of `amount` in Stripe functions. It is a string, not a number. We got away with this before because JS would implicitly convert for us. I verified this on a deployed contribution page.
- Changed z-index of react-alert so that they appeared above the modal overlay.

#### Why are we doing this? How does it help us?

Locks down state transitions on a contribution page so there are fewer ways things can go wrong.

#### How should this be manually tested? Please include detailed step-by-step instructions.

⚠️ This touches the logic for payment handling on contribution pages so careful QA is needed.

To test that validation that doesn't require a server roundtrip is working:
1. Go to a contribution page.
2. Complete everything except the Address field.
3. Click the Continue to Payment button.
4. Confirm you see an error message asking you to complete the field.
5. Repeat steps 1-4 for the City, State, and Country fields.

To test that validation that requires a server roundtrip works:
1. Go to a contribution page.
2. Complete everything except the First Name field.
3. Click the Continue to Payment button.
4. Confirm you see an error message asking you to complete the field.
5. Repeat steps 1-4 for the Last Name, Email, and Phone fields.

To test that finishing a one-time payment works:
1. Go to a contribution page.
2. Complete all fields, choosing a one-time payment.
3. Confirm that amount shown in the payment modal is accurate and has language around a one-time payment (e.g. mentions a single processing date).
4. Complete the payment.
5. Confirm the data you entered is accurate in Stripe and the dashboard.

To test that finishing a recurring payment works:
1. Go to a contribution page.
2. Complete all fields, choosing a monthly payment.
3. Confirm that amount shown in the payment modal is accurate and has language around a recurring payment (e.g. mentions that you will be billed on the correct date for the interval you selected).
4. Complete the payment.
5. Confirm the data you entered is accurate in Stripe and the dashboard.
6. Repeat steps 1-5 for a yearly payment.

To test that cancelling out of making a payment works:
1. Go to a contribution page.
6. Complete all fields.
7. Click Continue to Payment.
8. Click the Back button in the modal that appears.
9. Change all values in the form :(. Make sure to try changing the payment frequency, even.
10. Click Continue to Payment again.
11. Complete the payment.
12. Confirm that the payment shows correctly in Stripe and the dashboard.

To test that selecting amounts works:
1. Go to a contribution page.
2. Complete a contribution without clicking anything in the payment amount fields.
3. Confirm that the default payment amount shows correctly in Stripe and the dashboard.
4. Repeat steps 1-3, but click a preset amount button instead.
5. Repeat steps 1-3, but enter a custom amount instead.

To test that address line 2 works:
1. Go to a contribution page.
2. Complete a contribution, adding a second address line.
3. Confirm that the second address line shows in Stripe and the dashboard.

To test that the generic thank you redirection works:
1. Go to a contribution page that has no custom thank you page set.
2. Complete a contribution.
3. Confirm that the default thank you page appears, showing the correct amount.

To test that the thank you redirection works:
1. Go to a contribution page that has a custom thank you page set.
2. Complete a contribution.
3. Confirm that you are sent to the correct page after a contribution.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

n/a

#### Are there any smells or added technical debt to note?

No.

#### Has this been documented? If so, where?

No.

#### What are the relevant tickets? Add a link to any relevant ones.

https://news-revenue-hub.atlassian.net/browse/DEV-3619

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No.
